File: Samples/Digital signatures/SignWithPkcs11/VB.NET/Pkcs11Signer.vb
Imports System.Security.Cryptography.X509Certificates
Imports BitMiracle.Docotic.Pdf
Imports Net.Pkcs11Interop.Common
Imports Net.Pkcs11Interop.HighLevelAPI
Imports Net.Pkcs11Interop.HighLevelAPI.Factories

Namespace BitMiracle.Docotic.Pdf.Samples
    Class Pkcs11Signer
        Implements IPdfSigner
        Implements IDisposable

        Private ReadOnly m_session As ISession
        Private ReadOnly m_privateKey As IObjectHandle
        Private ReadOnly m_signatureAlgorithm As PdfSignatureAlgorithm
        Private ReadOnly m_digestAlgorithm As PdfDigestAlgorithm
        Private ReadOnly m_chain As X509Certificate2()
        Private m_disposed As Boolean

        Public Sub New(
            signatureAlgorithm As PdfSignatureAlgorithm,
            digestAlgorithm As PdfDigestAlgorithm,
            chain As X509Certificate2(),
            session As ISession,
            privateKey As IObjectHandle)

            m_signatureAlgorithm = signatureAlgorithm
            m_digestAlgorithm = digestAlgorithm
            m_chain = chain
            m_session = session
            m_privateKey = privateKey
        End Sub

        Public ReadOnly Property SignatureAlgorithm As PdfSignatureAlgorithm Implements IPdfSigner.SignatureAlgorithm
            Get
                Return m_signatureAlgorithm
            End Get
        End Property

        Public Function Sign(message As Byte()) As Byte() Implements IPdfSigner.Sign
            Dim mechanismFactory = New MechanismFactory()
            Dim mechanism As IMechanism
            If m_signatureAlgorithm = PdfSignatureAlgorithm.Ecdsa Then

                If m_digestAlgorithm = PdfDigestAlgorithm.Sha1 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_ECDSA_SHA1)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha256 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_ECDSA_SHA256)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha384 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_ECDSA_SHA384)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha512 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_ECDSA_SHA512)
                Else
                    Throw New InvalidOperationException("Unexpected digest algorithm " & m_digestAlgorithm)
                End If
            ElseIf m_signatureAlgorithm = PdfSignatureAlgorithm.Rsa Then

                If m_digestAlgorithm = PdfDigestAlgorithm.Sha1 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_SHA1_RSA_PKCS)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha256 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_SHA256_RSA_PKCS)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha384 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_SHA384_RSA_PKCS)
                ElseIf m_digestAlgorithm = PdfDigestAlgorithm.Sha512 Then
                    mechanism = mechanismFactory.Create(CKM.CKM_SHA512_RSA_PKCS)
                Else
                    Throw New InvalidOperationException("Unexpected digest algorithm " & m_digestAlgorithm)
                End If
            Else
                Throw New InvalidOperationException("Unexpected signature algorithm " & m_signatureAlgorithm)
            End If

            Return m_session.Sign(mechanism, m_privateKey, message)
        End Function

        Public Function GetChain() As X509Certificate2()
            Return m_chain
        End Function

        Public Sub Dispose() Implements IDisposable.Dispose
            If m_disposed Then Return
            m_disposed = True
            m_session.Dispose()

            For Each c In m_chain
                c.Dispose()
            Next

            GC.SuppressFinalize(Me)
        End Sub
    End Class
End Namespace

File: Samples/Digital signatures/SignWithPkcs11/C#/Pkcs11Signer.cs
using System;
using System.Security.Cryptography.X509Certificates;
using Net.Pkcs11Interop.Common;
using Net.Pkcs11Interop.HighLevelAPI;
using Net.Pkcs11Interop.HighLevelAPI.Factories;

namespace BitMiracle.Docotic.Pdf.Samples
{
    class Pkcs11Signer : IPdfSigner, IDisposable
    {
        private readonly ISession m_session;
        private readonly IObjectHandle m_privateKey;
        private readonly PdfSignatureAlgorithm m_signatureAlgorithm;
        private readonly PdfDigestAlgorithm m_digestAlgorithm;
        private readonly X509Certificate2[] m_chain;

        private bool m_disposed;

        public Pkcs11Signer(
            PdfSignatureAlgorithm signatureAlgorithm,
            PdfDigestAlgorithm digestAlgorithm,
            X509Certificate2[] chain,
            ISession session,
            IObjectHandle privateKey)
        {
            m_signatureAlgorithm = signatureAlgorithm;
            m_digestAlgorithm = digestAlgorithm;
            m_chain = chain;
            m_session = session;
            m_privateKey = privateKey;
        }

        public PdfSignatureAlgorithm SignatureAlgorithm => m_signatureAlgorithm;

        public byte[] Sign(byte[] message)
        {
            var mechanismFactory = new MechanismFactory();
            IMechanism mechanism = m_signatureAlgorithm switch
            {
                PdfSignatureAlgorithm.Ecdsa => m_digestAlgorithm switch
                {
                    PdfDigestAlgorithm.Sha1 => mechanismFactory.Create(CKM.CKM_ECDSA_SHA1),
                    PdfDigestAlgorithm.Sha256 => mechanismFactory.Create(CKM.CKM_ECDSA_SHA256),
                    PdfDigestAlgorithm.Sha384 => mechanismFactory.Create(CKM.CKM_ECDSA_SHA384),
                    PdfDigestAlgorithm.Sha512 => mechanismFactory.Create(CKM.CKM_ECDSA_SHA512),
                    _ => throw new InvalidOperationException("Unexpected digest algorithm " + m_digestAlgorithm),
                },
                PdfSignatureAlgorithm.Rsa => m_digestAlgorithm switch
                {
                    PdfDigestAlgorithm.Sha1 => mechanismFactory.Create(CKM.CKM_SHA1_RSA_PKCS),
                    PdfDigestAlgorithm.Sha256 => mechanismFactory.Create(CKM.CKM_SHA256_RSA_PKCS),
                    PdfDigestAlgorithm.Sha384 => mechanismFactory.Create(CKM.CKM_SHA384_RSA_PKCS),
                    PdfDigestAlgorithm.Sha512 => mechanismFactory.Create(CKM.CKM_SHA512_RSA_PKCS),
                    _ => throw new InvalidOperationException("Unexpected digest algorithm " + m_digestAlgorithm),
                },

                _ => throw new InvalidOperationException("Unexpected signature algorithm " + m_signatureAlgorithm),
            };
            return m_session.Sign(mechanism, m_privateKey, message);
        }

        public X509Certificate2[] GetChain() => m_chain;

        public void Dispose()
        {
            if (m_disposed)
                return;

            m_disposed = true;

            m_session.Dispose();

            foreach (var c in m_chain)
                c.Dispose();

            GC.SuppressFinalize(this);
        }
    }
}
